asset/compressed_file_reader/index.js

```diff
@@ -0,0 +1,210 @@
+'use strict';
+
+const Promise = require('bluebird');
+const fse = require('fs-extra');
+const os = require('os');
+const glob = require('glob');
+const _ = require('lodash');
+const { getOffsets, getChunk } = require('@terascope/chunked-file-reader');
+const { getOpConfig } = require('@terascope/job-components');
+
+async function newSlicer(context, job, retryData, logger) {
+    // Slicer queue.  Slices from multiple files can enqueue here but that's ok.
+    const slices = [];
+
+    // Limit concurrency, but do so in a way that one large upload will not hold
+    // up the entire batch the way Promise.map({concurrency}) would.
+    const working = [];
+
+    const opConfig = getOpConfig(job.config, 'compressed_file_reader');
+    const assetDir = await context.apis.assets.getPath('file-assets');
+
+    const filedb = await require('./filedb')(opConfig.workDir, assetDir);
+
+    async function sliceFile(src, work) {
+        const stat = await fse.stat(work);
+        const total = stat.size;
+        // NOTE: Important to update `slices` syncronously so that the correct
+        // `last` slice is marked.
+        getOffsets(opConfig.size, total, opConfig.delimiter).forEach((offset) => {
+            slices.push({
+                total,
+                work,
+                src,
+                ...offset,
+            });
+        });
+        // Mark the last slice so we know when to archive the file.
+        slices[slices.length - 1].last = true;
+        _.pull(working, src);
+        return work;
+    }
+
+    function doWork() {
+        if (filedb.numReady() === 0) {
+            return;
+        }
+        _.range(opConfig.concurrency - working.length).forEach(async () => {
+            const src = filedb.enqueue();
+            if (src) {
+                logger.info({ src }, 'decompressing');
+                working.push(src);
+                try {
+                    sliceFile(src, await filedb.decompress(src));
+                    logger.info({ src }, 'sliced & ready to read');
+                } catch (err) {
+                    // Not sure how to handle. Retry with exponential backoff before giving up?
+                    _.pull(working, src);
+                    logger.error({ stderr: err.stderr }, err);
```

Comment:
I figured better to wait until get more experience handling failure after more testing.

Reply:
Would you just want to fail the slice and let the framework handle the retry?

Reply:
As we discussed, since this is in the slicer, there isn't any retry at this level.  I'd like to leave this failing early & hard until we get some more operational insight into the kinds of errors we'll see and to what extent retrying could help.  The only thing that comes to mind that could benefit is if the volume ran out of space.